diff --git a/gdk/quartz/GdkQuartzWindow.c b/gdk/quartz/GdkQuartzWindow.c
index dabf051..5963660 100644
--- a/gdk/quartz/GdkQuartzWindow.c
+++ b/gdk/quartz/GdkQuartzWindow.c
@@ -617,7 +617,6 @@ update_context_from_dragging_info (id <NSDraggingInfo> sender)
 - (void)draggedImage:(NSImage *)anImage endedAt:(NSPoint)aPoint operation:(NSDragOperation)operation
 {
   GdkEvent event;
-  GdkScreen *screen;
 
   g_assert (_gdk_quartz_drag_source_context != NULL);
 
@@ -626,7 +625,7 @@ update_context_from_dragging_info (id <NSDraggingInfo> sender)
   event.dnd.send_event = FALSE;
   event.dnd.context = _gdk_quartz_drag_source_context;
 
-  screen = gdk_window_get_screen (event.dnd.window);
+  GdkScreen* screen = gdk_window_get_screen (event.dnd.window);
 
   if (screen)
     {
@@ -649,6 +648,7 @@ update_context_from_dragging_info (id <NSDraggingInfo> sender)
           wh = gdk_window_get_height (win);
 
           if (gx > wx && gy > wy && gx <= wx + ww && gy <= wy + wh)
+		  /* found a toplevel GdkWindow at the drop position */
             event.dnd.context->dest_window = win;
         }
     }
diff --git a/gtk/gtkdnd-quartz.c b/gtk/gtkdnd-quartz.c
index 62b8570..74bdbc7 100644
--- a/gtk/gtkdnd-quartz.c
+++ b/gtk/gtkdnd-quartz.c
@@ -1926,7 +1926,11 @@ gtk_drag_source_info_destroy (GtkDragSourceInfo *info)
 }
 
 static gboolean
-drag_drop_finished_idle_cb (gpointer data)
+    GtkDragSourceInfo* info = (GtkDragSourceInfo*) data;
+    if (info->success) 
+      {
+         gtk_drag_source_info_destroy (data);
+      }
 {
   GtkDragSourceInfo* info = (GtkDragSourceInfo*) data;
 
